File: 3D Touch测试/3D Touch测试/ViewController.m
//
//  ViewController.m
//  3D Touch测试
//
//  Created by apple on 2017/3/7.
//  Copyright © 2017年 apple. All rights reserved.
//

#import "ViewController.h"
#import "DetailViewController.h"

@interface ViewController ()<UITableViewDelegate, UITableViewDataSource, UIViewControllerPreviewingDelegate>
/*tableView*/
@property(nonatomic, strong)UITableView * tableView;
@end

@implementation ViewController

- (UITableView *)tableView {
    if (!_tableView) {
        _tableView = ({
            UITableView *object = [[UITableView alloc] initWithFrame:self.view.bounds style:(UITableViewStyleGrouped)];
            object.delegate= self;
            object.dataSource = self;
            object;
        });
    }
    return _tableView;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setup];
}

- (void)setup {
    [self.view addSubview:self.tableView];
//    if ([[UIDevice currentDevice].systemVersion floatValue] >= 9.0) {
//        if (self.traitCollection.forceTouchCapability == UIForceTouchCapabilityAvailable) {
//            [self registerForPreviewingWithDelegate:self sourceView:self.tableView];
//        }
//    }
}
#pragma - mark UIViewControllerPreviewingDelegate
- (nullable UIViewController *)previewingContext:(id <UIViewControllerPreviewing>)previewingContext viewControllerForLocation:(CGPoint)location NS_AVAILABLE_IOS(9_0) {
    DetailViewController *detailVC = [[DetailViewController alloc] init];
    detailVC.preferredContentSize = CGSizeMake(200, 200);
    detailVC.view.backgroundColor = [UIColor blueColor];
    //[previewingContext sourceView]拿到对应的cell；
    NSIndexPath *indexPath = [_tableView indexPathForCell:(UITableViewCell* )[previewingContext sourceView]];
    NSLog(@"%@", indexPath);
    return detailVC;
}
- (void)previewingContext:(id <UIViewControllerPreviewing>)previewingContext commitViewController:(UIViewController *)viewControllerToCommit NS_AVAILABLE_IOS(9_0) {
    [self.navigationController pushViewController:viewControllerToCommit animated:true];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}
#pragma - mark UITableViewDelegate UITableViewDataSource
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 3;
}
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 3;
}
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell"];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:(UITableViewCellStyleValue2) reuseIdentifier:@"cell"];
    }
    cell.textLabel.text = [NSString stringWithFormat:@"%ld-%ld", (long)indexPath.section, (long)indexPath.row];
    //注册peek
    if ([[UIDevice currentDevice].systemVersion floatValue] >= 9.0) {
        if (self.traitCollection.forceTouchCapability == UIForceTouchCapabilityAvailable) {
            [self registerForPreviewingWithDelegate:self sourceView:cell];
        }
    }

    return cell;
}
@end

File: Swift练习/Swift iOS应用开发实战/04-Storyboard组织和管理视图/4.1__Shopping/4.1__Shopping/ViewController.swift
//
//  ViewController.swift
//  4.1__Shopping
//
//  Created by M-coppco on 16/6/19.
//  Copyright © 2016年 M-coppco. All rights reserved.
//

import UIKit

//遵循协议
class ViewController: UIViewController, UITableViewDelegate, UITableViewDataSource, NewItemViewControllerDelegate {
    @IBOutlet weak var tableView: UITableView!

    var buyItems = [
        HJItem.init(itemName: "牛奶", itemBrand: "三元"),
        HJItem(itemName: "红烧牛肉面", itemBrand: "康师傅"),
        HJItem(itemName: "桃汁", itemBrand: "汇源"),
        HJItem(itemName: "巧克力", itemBrand: "德芙"),
        HJItem(itemName: "地板净", itemBrand: "滴露"),
        HJItem(itemName: "洗发水", itemBrand: "飘柔")
    ]
    /*
    let buyItems = [("牛奶", "三元"),
        ("红烧牛肉面", "康师傅"),
        ("桃汁", "汇源"),
        ("巧克力", "德芙"),
    ("地板净", "滴露"),
        ("洗发水", "飘柔")]
*/
    override func viewDidLoad() {
        super.viewDidLoad()
        //iOS 8需要添加这句
        self.edgesForExtendedLayout = UIRectEdge.None
        //去掉多余分割线
        tableView.tableFooterView = UIView()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func viewWillAppear(animated: Bool) {
        tableView.reloadData()
    }

    //执行segue的时候 前面执行
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "showItem" {
            let destinationVC : ItemViewController = segue.destinationViewController as! ItemViewController
            let item = buyItems[sender as! NSInteger]
            destinationVC.item = item
        } else if segue.identifier == "newItem" {
            let destinationVC : NewItemViewController = segue.destinationViewController as! NewItemViewController
            destinationVC.delegate = self
        }
    }
    
    func addNewItem(controller: NewItemViewController, item: HJItem) {
        buyItems.append(item)
        tableView.reloadData()
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return buyItems.count;
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("cell", forIndexPath: indexPath)
        let item = buyItems[indexPath.row]
        cell.textLabel?.text = item.itemName
        cell.detailTextLabel?.text = item.itemBrand
        if item.isBuy {
            cell.textLabel?.textColor = UIColor.greenColor()
        } else {
            cell.textLabel?.textColor = UIColor.redColor()
        }
        return cell
    }
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        //let (item, brand) = buyItems[indexPath.row]
        self.performSegueWithIdentifier("showItem", sender: indexPath.row)
        tableView.deselectRowAtIndexPath(indexPath, animated: true)
    }
   
}

